daffodil-propgen/src/main/resources/org/apache/daffodil/xsd/dafext.xsd

```diff
@@ -721,6 +721,7 @@
           <xs:enumeration value="layerCompileWarning" />
           <xs:enumeration value="multipleChoiceBranches" />
           <xs:enumeration value="namespaceDifferencesOnly" />
+          <xs:enumeration value="nodeIndexedLikeArray" />
```

Comment:
Note that the current check only detects when self expression is indexed like an array. It doesn't detect all single nodes indexed like an array. For example, you could have an expression like `foo/bar/baz[1]`, where baz is not an array. In that case we probably also want a similar `nodeIndexLikeArray` warning. We might want to also to detect those cases in this PR if it's similar enough, could be done separately though. I believe Expression.scala already knows which steps are arrays so it might be as simple as
```scala
if (!step.isArray && predArg.isDefined) SDW(...)
```

Reply:
Steve makes a good point. If you have elements and arrays, and you change the nesting, and in `foo/bar/baz[1]` baz was an array before, but now it's an enclosing scalar element or inside the array items, but it's not an array anymore, something else is, then, the expression wants to be broken (or at least warn) because you're not indexing the array anymore, you're indexing in a no-op manner, a scalar. 

I've actually made exactly this error, and had to search all over the place for where I was indexing the wrong element, having introduced a level of element someplace.

Reply:
Does this want to be an error or remain a warning? I found the following in the DFDL spec under Section 35 under Expression Errors. So should we be SDEing here if step is not an array or an optional element?

```
 Indexing of non-array non-optional element

§  Example: x[1] when x is declared and has both XSD minOccurs="1" and XSD maxOccurs="1" explicitly, or by not stating either or both of them.
```

Reply:
It looks like we actually might already have the SDE check somewhere? If I create an optional element and try to index it with `[1]` I get the SDE:

> [error] Schema Definition Error: Subset: Indexing is only allowed on arrays. Offending path step: 'foo[1]'.

So somewhere we make sure you can only have a predicate on arrays, and we go one step further and say you can't even have it on optionals. I think maybe allowing it on optionals was mentioned in the spec because Daffodil used to treat optionals as arrays, but I don't think that's the case anymore?

If we are going to be consistent with that behavior, I think maybe `.[1]` and `..[1]` want to be changed to SDE's also, since self-step and up-step can never reference an whole array, they always reference a single element (which at best could be a single element of an array).

I grepped through the regression suite and I don't see any schemas that do something like .[1], so I don't think this is something we need to support. Maybe at some point it was needed, but that doesn't seem to be the case anymore.

Reply:
Agree. Let's change the goal of this ticket and make them SDEs. This notion that a DFDL implementation can just use XPath is a fiction that was never realistic, and it will be entirely moot once there is a non-XSD syntax for DFDL (which is planned for DFDL v2.0)

XPath was designed to work without schemas, as was almost everything about XML. In a strongly typed world with schemas, you don't need to have this much squishyness about what's meaningful, and what isn't.